working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+}
+
+int _nextId = 0;
+final _cache = <int, C>{};
+
+C makeObject() {
+  _cache[nextId] = C(nextId);
+  nextId++;
+  return _cache[nextId - 1];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the
+same code becomes thread _unsafe_ in the presence of shared memory
+multithreading.
+
+### Generics
+
+When declaring a generic type the developer will have to use type parameter
+bounds to ensure that resulting class conforms to restrictions imposed by
+`Shareable`:
+
+```dart
+class X<T> implements Shareable {
+  T v;  // compile time error.
+}
+
+class Y<T extends Shareable> implements Shareable {
+  T v;  // ok.
+}
+```
+
+> [!NOTE]
+>
+> We could really benefit from the ability to use intersection types here, which
+> would allow to specify complicated bounds like `T extends Shareable & I`. In
+> the absence of intersections types developers would be forced to declare
+> intermediate interfaces which implement all required interfaces (e.g.
+> `abstract interface ShareableI implements Shareable, I {}`) and require users
+> to implement those by specifying `T extends ShareableI`.
+
+### Functions
+
+Shareability of a function depends on the values that it captures. We could
+define that any **function is shareable iff it captures only variables of
+shareable type**. Incorporating this property into the type system naturally
+leads to the desire to use _intersection types_ to express the property that
+some value is both a function of a specific type _and_ shareable:
+
+```dart
+class A implements Shareable {
+  void Function() f;  // compile time error
+  void Function() & Shareable f;  // ok
+}
+```
+
+Introducing intersection types into type system might be a huge undertaking. For
+the purposes of developing an MVP we can choose one of the two approaches:
+
+- Ignore functions entirely: consider functions un-shareable. It becomes a
+  compile time error to have a function type field inside a shareable class.
+- Allow function type fields inside a shareable class, but enforce shareability
+  in runtime on assignment to the field.
+
+### Shareable core types
+
+The following types will become shareable:
+
+- `num` (`int` and `double`), `String`, `bool`, `Null`, `BigInt`
+- `Enum` and consequently all user defined enums
+- `RegExp`, `DateTime`, `Uri`
+- `TypedData` - which makes all typed data types shareable.
+- `Pointer`
+- `Type` and `Symbol`
+- `StackTrace`
+
+#### Collections
+
+`dart:core` will provide shareable variants of all collection classes:
+
+```dart
+abstract interface class ShareableList<E extends Shareable?>
+    implements List<E>, Shareable {
+}
+
+abstract interface class ShareableSet<E extends Shareable?>
+    implements Set<E>, Shareable {
+}
+
+abstract interface class ShareableMap<K extends Shareable?,
+                                  V extends Shareable?>
+    implements Map<K, V>, Shareable {
+}
+```
+
+Default implementations of `List`, `Set` and `Map` will be changed to be
```

Comment:
Are you suggesting this only for literals, or also for constructors?
And would this be special cased magic in the type checker in the CFE and analyzer to promote the type from List to ShareableList?

If we had overloading of factory constructors, and selected the most precise type based on type argument bounds, we wouldn't have to special case it.

Reply:
> Are you suggesting this only for literals, or also for constructors?

Constructors will dynamically return the shareable type based on the concrete value of the type parameter. 

> And would this be special cased magic in the type checker in the CFE and analyzer to promote the type from List to ShareableList?

I think we would want to avoid doing anything to static types because it muddies the story considerably. @lrhn has given some examples where strange things start happening in type inference , e.g. `v ? <Shareable>[] : <Object>[]` yields `Object` and not `List<Object>` as one would think (because we are trying to figure out LUB of `ShareableList<Shareable>()` and `List<Object>()` and the current algorithm apparently can't deal with this).

> If we had overloading of factory constructors, and selected the most precise type based on type argument bounds, we wouldn't have to special case it.

We also want things to work in the case when `T` is not statically known, e.g. 

```dart
List<T> someGenericMethod<T>(int num, T initialValue) {
  final list = List<T>.filled(num, initialValue);
  // if T is Shareable we would like list to also be Shareable.
}
```